pubtools/exodus/hooks/pulp.py

```diff
@@ -0,0 +1,165 @@
+import logging
+import os
+import sys
+
+import backoff
+import requests
+from pubtools.pluggy import hookimpl, pm  # pylint: disable=wrong-import-order
+from six.moves.urllib.parse import urljoin
+
+LOG = logging.getLogger("pubtools-exodus")
+
+
+class ExodusPulpHanlder:
+    def __init__(self):
+        self._gw_url = os.getenv("EXODUS_GW_URL")
+        self._gw_env = os.getenv("EXODUS_GW_ENV")
+        self._gw_crt = os.getenv("EXODUS_GW_CERT")
+        self._gw_key = os.getenv("EXODUS_GW_KEY")
+
+        self._publish = None
+        self._check_cert()
+
+    def _new_session(self):
+        out = requests.Session()
+        out.cert = (self._gw_crt, self._gw_key)
+        return out
+
+    def _do_request(self, **kwargs):
+        session = self._new_session()
+        resp = session.request(**kwargs)
+        self._unpack_response(resp)
+        return resp
+
+    def _check_cert(self):
+        """Issue request to exodus-gw to identify permissions."""
+
+        if not self._gw_crt:
+            LOG.debug("EXODUS_GW_CERT not found, authentication may fail")
+
+        if not self._gw_key:
+            LOG.debug("EXODUS_GW_KEY not found, authentication may fail")
+
+        auth_url = urljoin(self._gw_url, "/whoami")
+        resp = self._do_request(method="GET", url=auth_url)
+        context = resp.json()
+
+        for user_type, ident in (
+            ("client", "serviceAccountId"),
+            ("user", "internalUsername"),
+        ):
+            typed_ctx = context[user_type]
+            if typed_ctx["authenticated"]:
+                LOG.debug(
+                    "You are authenticated as %s %s with roles: %s",
+                    user_type,
+                    typed_ctx[ident],
+                    [str(role) for role in typed_ctx["roles"]],
+                )
+                break
+        else:
+            LOG.debug("You are not authenticated with exodus-gw.")
+
+    @classmethod
+    def _unpack_response(cls, response):
+        """Raise if 'response' was not successful.
+
+        This is the same as response.raise_for_status(), merely wrapping it
+        to ensure the body is logged when possible."""
+
+        try:
+            response.raise_for_status()
+        except Exception as outer:
+            try:
+                body = response.json()
+            except:
+                raise outer
+
+            LOG.error("Unsuccessful response from exodus-gw: %s", body)
+            raise
+
+    @backoff.on_predicate(
```

Comment:
The goal is for this code to be used from Pub - have you checked if 'backoff' module is readily available there?  i.e. is it available as an eng-rhel-6 package and is there even a version which works on py2.6.

If it's not already available, you probably want to just drop this as making it available will be too much additional work.